### ISSUE TITLE
Add stateful processing

### ISSUE THREAD

**Issue by gdt** [merbanan/rtl_433#2635]
It has become clear that some stateful processing is needed.  rtl_433 itself by doctrine emits one message for each received transmission.   Straightforward processing means that for say 3 repeats of a temperature sensor (done by the device for reliability), one gets 3 mqtt messages or 3 influxdb entries.   This is particularly important for events.  What's needed is an approach to coalesce these, which could be:
  - something that receives syslog/json and emits a reduced stream, to use as a pipe component between rtl_433 and rtl_433_mqtt_relay
  - some other way to plumb that
  - built into rtl_433_mqtt_relay, but this is harder to reuse, so perhaps a python module that can be used in various programs

**Comment by gdt** [comment]
See #908 for filtering in rtl_433, but this seems counter to doctrine.

**Comment by merbanan** [comment]
I think state-full and duplicates are different things. The dup-check should filter things that belong to signal redundancy. The state-full logic should handle cases when several signals belong to the same message.

**Comment by merbanan** [comment]
https://github.com/merbanan/rtl_433/pull/2314 contains a suggestion from me on an implementation for state-full support in the decoders.

**Comment by rct** [comment]
Here's another use case (or two) that I think is worth considering:

With a reasonably large number of sensors, particularly those that have a volatile component in their ID, it is useful to have easy access to some meta data about devices, examples include:

- When a device was first heard  
- When a device was last heard
- When battery state last changed

These would be more useful if they were retained across rtl_433 invocations to be more useful. 

Why: 

- an existing known device might have intentionally or unintentionally been power cycled and now have a new ID that makes it essentially a new device. 
- If multiple devices of the same type are in use, it can be difficult to figure out which device is which. 

Note: there are some statistics tracked by rtl_433, but in most cases they aren't that useful for managing an environment with a bunch of devices. The stats that are available are geared more towards decoder development.  Statistics/meta data that would be useful for managing the environment often require tracking state. 

Architecturally, I'd like to see rtl_433 do the best job it can managing the radio and presenting the bits in ways other applications can easily consume.  The flex decoders are a great example that seem to make this case.  There are some devices where the output of the flex decoder is sufficient. For others, something needs parse the output and turn it into what ever the use case calls for. 

So in line with what @gdt was mentioning at the start of this thread, a way to address these needs, is to have some sort of framework that handles all of the boilerplate scaffolding for receiving messages, using timers, and then being able to send the generated data/messages to the desired destinations.  It should be easy to implement a piece of code to get the desired applicaiton logic without having to re-implement all of the scaffolding.

A number of the example apps (mqtt relay, hass autodiscovery, ...) could benefit from being implemented in a common framework with common scaffolding, that has the full event loop type semantics. 

I think this approach could handle many cases. I've skimmed #2314 but I don't understand whether what's needed there could be implemented in a consumer or really needs to be in rtl_433 for some reason.

**Comment by gdt** [comment]
I agree that duplicate suppression could be separated from this.  I didn't write it this way because I have perceived that there is doctrine against putting this in base rtl_433.    My own view is that duplicates are almost entirely unhelpful (unless one is assessing signal delivery, in which case a "don't do dup suppression" flag is in order).    I'm happy to add an issue for dups in rtl_433 proper, post decode and pre emission of all types, and let this be about more complicated processing.

**Comment by gdt** [comment]
For rct@'s example, I see a "keep track of history" as a database/consumer sort of parallel to HA rather than a core rtl_433 feature.    I'm really thinking about transforming "message received" into sensible semantics for some other system to process, and I really started down this mental path (more than just dups) because of Govee.

**Comment by gdt** [comment]
> #2314 contains a suggestion from me on an implementation for state-full support in the decoders.

That seems to be about what I might call MAC-level fragmentation, where two data bursts are one logical packet and it's not possible to even do CRC checking on them separately.   It sounds like "just emit json with the bits from each and let some other program deal" isn't going to work.    I tend to think we should limit stateful processing to those sorts of cases (as irregular and otherwise intractable), and use a higher-level approach for things that are more normal, like "if it isn't reporting wet, and we heard from it, it is now dry".

**Comment by merbanan** [comment]
What I aim for is to design robust solutions. In the case where a device uses random ids there is no way to make it unique in a stable way by only using the id. In theory you could separate the devices by figuring out the cycle time and RSSI. But I dont think we can assume it will be linear over time. So the only way is to look at some other parameter and that is the channel if it is possible to change it. That way you often have the ability to use 3 sensors of that model.

So in this case the sensor has its limits and I dont think we should bend over backwards to give the illusion that it is any better.

What @gdt write is what I feel could be in the scope of rtl_433. MAC-level fragmentation and duplicate filtering. More advanced thing that would require keeping state longer and using timers does not really fit with the current error handling logic (just exit and restarting).

**Comment by merbanan** [comment]
But in the case of emitting implied state from a sensor that would be ok if it makes it easier to use. But I have devices which are hard to use. It is a contact sensor that only triggers when the contact is open but not when it is closed. Using that sensor in any kind of system needs something to keep state.

**Comment by gdt** [comment]
I completely understand about trying to guess ids being bad.  I have split off the two specific things we agree can go in core rtl_433 into #2640 and #2641.  This issue is now about adding some sort of middleware between json-format deocdes and sending to HA etc. via mqtt or some similar mechanisms.   A candidate architecture is a python library that accepts json objects  and some kind of timer tick, and produces json objects, that one can interpose between rtl_433 and sending to data consumers.

**Comment by gdt** [comment]
Another use case, discussed in #2540,  is to map ids (from badly-behaved devices that change their id on powerup) into fixed tokens, so that later processing (db, HA, whatever) can operate on logical devices.   Per doctrine, this doesn't belong in rtl_433 core.

**Comment by sheilbronn** [comment]
In the recent years I built a bash wrapper daemon around rtl_433 to cover some of the use cases mentioned here.
You might want to have a look at its feature list for inspirations: [rtl2mqtt](https://github.com/sheilbronn/rtl2mqtt).
What might be interesting here:
* HA MQTT discovery (not immediately after the first message to avoid errors)
* dewpoint approximation
* configurable rounding of values (less flicker)
* ... other stuff

**Comment by gdt** [comment]
Another use case, discussed in #2105, is to filter bad data by some recognition of values that are not physically possible (temp changes at super fast timescales).

**Comment by merbanan** [comment]
I think we have to keep true to the unofficial mission and that is to decode rf data as clean as possible. What might be an option is a selectable X seconds duplicate filtering step. 

Regarding the few state-full decoders we have I have an idea of storing the needed state outside of the decoders. That way we get rid of state in the decoders and we can have many demodulators calling the same decoder.

**Comment by sheilbronn** [comment]
Thanks for the valuable discussion this subject of deduplication. 
Especially some motion sensors send data > 6-times to increase probability of reception. Also some outdoor sensors send temperature data every 30 seconds, which both clutters log files and fires MQTT events to frequently!
(This was one the motivations I also built early sensor specific deduplification into my rtl2mqtt wrapper script: https://github.com/sheilbronn/rtl2mqtt)

**Comment by merbanan** [comment]
Here we have the most recent discussion: https://github.com/merbanan/rtl_433/issues/3620

But to answer you the motion sensor should be possible to handle but the 30 second sensor reading would not be handled as it is a truly separate event.

**Comment by merbanan** [comment]
What would be of interest is what sensors get unnecessary repeats from.

**Comment by sheilbronn** [comment]
> Here we have the most recent discussion: [#3620](https://github.com/merbanan/rtl_433/issues/3620)
> 
> But to answer you the motion sensor should be possible to handle but the 30 second sensor reading would not be handled as it is a truly separate event.

Totally agree, just wanted to mention it for clarity.

**Comment by sheilbronn** [comment]
> What would be of interest is what sensors get unnecessary repeats from.

In my case my old Sonoff motion sensors report motion events with approx. 10 repeated messages like:

`{"time":"2026-07-19 14:21:58","protocol":30,"model":"Generic-Remote","id":61825,"cmd":62,"tristate":"110ZX00Z011X","mod":"ASK","freq":433.952,"rssi":1.96,"snr":14.391,"noise":-14.392}`

My code removes all motion events approx. 2 seconds within the first reported event from the sensors.

(Will also continue discussion in #3620 from now on.)
